File: apps/api/src/bimstitch_api/schemas/bcf.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Viewpoint
# ---------------------------------------------------------------------------


class Vec3Schema(BaseModel):
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0


class BcfComponentsSchema(BaseModel):
    default_visibility: bool = True
    visibility_exceptions: list[str] = Field(default_factory=list)
    selection: list[str] = Field(default_factory=list)


class ClippingPlaneSchema(BaseModel):
    location: Vec3Schema
    direction: Vec3Schema


class XrayOpacityOverrideSchema(BaseModel):
    global_id: str
    opacity: float


class XrayStateSchema(BaseModel):
    items: list[str] = Field(default_factory=list)
    opacity_overrides: list[XrayOpacityOverrideSchema] = Field(default_factory=list)


class BcfMeasurementSchema(BaseModel):
    type: str
    points: list[Vec3Schema] = Field(default_factory=list)
    height: float | None = None


class ViewState2DSchema(BaseModel):
    center_x: float = 0.0
    center_y: float = 0.0
    zoom: float = 1.0
    visible_layers: list[str] = Field(default_factory=list)
    file_type: str = "dxf"
    # PDF markup extension. `page` is the 1-based page; `annotations` holds the
    # markup shapes (kept as free dicts — they are app-private and ignored by the
    # standard .bcf export). See the viewer's `Annotation2D` type.
    page: int | None = None
    annotations: list[dict[str, Any]] = Field(default_factory=list)


class BcfViewpointCreate(BaseModel):
    guid: str = Field(max_length=36)
    index_in_topic: int = 0
    camera_type: str = Field(max_length=20)
    camera_view_point: Vec3Schema
    camera_direction: Vec3Schema
    camera_up_vector: Vec3Schema
    field_of_view: float | None = None
    field_of_height: float | None = None
    components: BcfComponentsSchema | None = None
    clipping_planes: list[ClippingPlaneSchema] = Field(default_factory=list)
    xray: XrayStateSchema | None = None
    measurements: list[BcfMeasurementSchema] = Field(default_factory=list)
    is_2d: bool = False
    view_state_2d: ViewState2DSchema | None = None
    linked_file_id: UUID | None = None


class BcfViewpointRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    guid: str
    index_in_topic: int
    camera_type: str
    camera_view_point: dict[str, float]
    camera_direction: dict[str, float]
    camera_up_vector: dict[str, float]
    field_of_view: float | None
    field_of_height: float | None
    components: dict | None
    clipping_planes: list | None
    xray: dict[str, Any] | None = None
    measurements: list[Any] | None = None
    snapshot_url: str | None = None
    is_2d: bool
    view_state_2d: dict | None
    linked_file_id: UUID | None
    created_at: datetime


# ---------------------------------------------------------------------------
# Comment
# ---------------------------------------------------------------------------


class BcfCommentCreate(BaseModel):
    text: str = Field(min_length=1, max_length=4000)
    viewpoint_guid: str | None = Field(default=None, max_length=36)


class BcfCommentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    guid: str
    comment_text: str
    author: str
    date: datetime
    modified_author: str | None
    modified_date: datetime | None
    viewpoint_guid: str | None
    created_by_user_id: UUID | None
    created_at: datetime


# ---------------------------------------------------------------------------
# Topic
# ---------------------------------------------------------------------------


class BcfTopicCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    topic_type: str = Field(default="Issue", max_length=50)
    topic_status: str = Field(default="Open", max_length=50)
    priority: str | None = Field(default=None, max_length=50)
    stage: str | None = Field(default=None, max_length=50)
    assigned_to: str | None = Field(default=None, max_length=255)
    labels: list[str] = Field(default_factory=list)
    due_date: date | None = None
    linked_finding_id: UUID | None = None
    linked_model_id: UUID | None = None
    # The specific model version (ProjectFile) the issue was raised against, and
    # whether this is a 2D (drawing) or 3D (IFC) issue. Both auto-derive from the
    # initial viewpoint when omitted (see routers/bcf.py::create_topic).
    linked_file_id: UUID | None = None
    is_2d: bool = False
    viewpoint: BcfViewpointCreate | None = None


class BcfTopicUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    topic_type: str | None = Field(default=None, max_length=50)
    topic_status: str | None = Field(default=None, max_length=50)
    priority: str | None = Field(default=None, max_length=50)
    stage: str | None = Field(default=None, max_length=50)
    assigned_to: str | None = Field(default=None, max_length=255)
    labels: list[str] | None = None
    due_date: date | None = None
    linked_finding_id: UUID | None = None
    linked_model_id: UUID | None = None
    linked_file_id: UUID | None = None
    is_2d: bool | None = None


class BcfTopicRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    guid: str
    title: str
    description: str | None
    topic_type: str
    topic_status: str
    priority: str | None
    stage: str | None
    assigned_to: str | None
    labels: list[str] | None
    due_date: date | None
    creation_author: str
    creation_date: datetime
    modified_author: str | None
    modified_date: datetime | None
    linked_finding_id: UUID | None
    linked_model_id: UUID | None
    linked_file_id: UUID | None
    is_2d: bool
    # Derived from the linked ProjectFile (version raised against) for display.
    model_version: int | None = None
    file_type: str | None = None
    created_by_user_id: UUID
    bcf_version: str
    import_source: str | None
    created_at: datetime
    updated_at: datetime

    viewpoints: list[BcfViewpointRead] = Field(default_factory=list)
    comments: list[BcfCommentRead] = Field(default_factory=list)


class BcfTopicSummary(BaseModel):
    """Lightweight read model for topic list (no nested viewpoints/comments)."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    guid: str
    title: str
    topic_type: str
    topic_status: str
    priority: str | None
    assigned_to: str | None
    creation_author: str
    creation_date: datetime
    linked_finding_id: UUID | None
    linked_model_id: UUID | None = None
    linked_file_id: UUID | None = None
    is_2d: bool = False
    model_version: int | None = None
    file_type: str | None = None
    # Whether the topic has at least one viewpoint to jump to. Drives the
    # "go to viewpoint" button — independent of whether a snapshot image exists.
    has_viewpoint: bool = False
    snapshot_url: str | None = None
    created_at: datetime


class BcfMarkup2DItem(BaseModel):
    """A 2D markup topic projected for rendering on a PDF page.

    Returned by ``GET /bcf-topics/markup-2d?file_id=`` — one entry per topic
    whose viewpoint is 2D and linked to the given file. ``annotations`` is the
    raw markup-shape list stored in the viewpoint's ``view_state_2d``.
    """

    topic_id: UUID
    title: str
    topic_status: str
    page: int | None
    annotations: list[dict[str, Any]] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Import / Export
# ---------------------------------------------------------------------------


class BcfImportResponse(BaseModel):
    imported_count: int
    topics: list[BcfTopicRead]
    warnings: list[str] = Field(default_factory=list)


class BcfExportResponse(BaseModel):
    download_url: str
    filename: str


class BcfLinkFindingRequest(BaseModel):
    finding_id: UUID


class BcfFromFindingsRequest(BaseModel):
    finding_ids: list[UUID] = Field(min_length=1)

File: apps/portal/src/lib/api/schemas/bcf.ts
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Viewpoint
// ---------------------------------------------------------------------------

export const BcfVec3Schema = z.object({
  x: z.number(),
  y: z.number(),
  z: z.number(),
});

export const BcfComponentsSchema = z.object({
  default_visibility: z.boolean(),
  visibility_exceptions: z.array(z.string()),
  selection: z.array(z.string()),
});

export const BcfClippingPlaneSchema = z.object({
  location: BcfVec3Schema,
  direction: BcfVec3Schema,
});

export const BcfViewpointReadSchema = z.object({
  id: z.string().uuid(),
  guid: z.string(),
  index_in_topic: z.number(),
  camera_type: z.string(),
  camera_view_point: z.record(z.string(), z.number()),
  camera_direction: z.record(z.string(), z.number()),
  camera_up_vector: z.record(z.string(), z.number()),
  field_of_view: z.union([z.number(), z.null()]),
  field_of_height: z.union([z.number(), z.null()]),
  components: z.union([z.record(z.string(), z.unknown()), z.null()]),
  clipping_planes: z.union([z.array(z.unknown()), z.null()]),
  snapshot_url: z.union([z.string(), z.null()]).optional(),
  is_2d: z.boolean(),
  view_state_2d: z.union([z.record(z.string(), z.unknown()), z.null()]),
  linked_file_id: z.union([z.string().uuid(), z.null()]),
  created_at: z.string(),
});

export type BcfViewpointRead = z.infer<typeof BcfViewpointReadSchema>;

// ---------------------------------------------------------------------------
// Comment
// ---------------------------------------------------------------------------

export const BcfCommentReadSchema = z.object({
  id: z.string().uuid(),
  guid: z.string(),
  comment_text: z.string(),
  author: z.string(),
  date: z.string(),
  modified_author: z.union([z.string(), z.null()]),
  modified_date: z.union([z.string(), z.null()]),
  viewpoint_guid: z.union([z.string(), z.null()]),
  created_by_user_id: z.union([z.string().uuid(), z.null()]),
  created_at: z.string(),
});

export type BcfCommentRead = z.infer<typeof BcfCommentReadSchema>;

// ---------------------------------------------------------------------------
// Topic
// ---------------------------------------------------------------------------

export const BcfTopicReadSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid(),
  guid: z.string(),
  title: z.string(),
  description: z.union([z.string(), z.null()]),
  topic_type: z.string(),
  topic_status: z.string(),
  priority: z.union([z.string(), z.null()]),
  stage: z.union([z.string(), z.null()]),
  assigned_to: z.union([z.string(), z.null()]),
  labels: z.union([z.array(z.string()), z.null()]),
  due_date: z.union([z.string(), z.null()]),
  creation_author: z.string(),
  creation_date: z.string(),
  modified_author: z.union([z.string(), z.null()]),
  modified_date: z.union([z.string(), z.null()]),
  linked_finding_id: z.union([z.string().uuid(), z.null()]),
  linked_model_id: z.union([z.string().uuid(), z.null()]),
  created_by_user_id: z.string().uuid(),
  bcf_version: z.string(),
  import_source: z.union([z.string(), z.null()]),
  created_at: z.string(),
  updated_at: z.string(),
  viewpoints: z.array(BcfViewpointReadSchema),
  comments: z.array(BcfCommentReadSchema),
});

export type BcfTopicRead = z.infer<typeof BcfTopicReadSchema>;

export const BcfTopicSummarySchema = z.object({
  id: z.string().uuid(),
  guid: z.string(),
  title: z.string(),
  topic_type: z.string(),
  topic_status: z.string(),
  priority: z.union([z.string(), z.null()]),
  assigned_to: z.union([z.string(), z.null()]),
  creation_author: z.string(),
  creation_date: z.string(),
  linked_finding_id: z.union([z.string().uuid(), z.null()]),
  snapshot_url: z.union([z.string(), z.null()]).optional(),
  created_at: z.string(),
});

export type BcfTopicSummary = z.infer<typeof BcfTopicSummarySchema>;

export const BcfTopicListSchema = z.array(BcfTopicSummarySchema);

export type BcfTopicList = z.infer<typeof BcfTopicListSchema>;

// ---------------------------------------------------------------------------
// Create / Update inputs
// ---------------------------------------------------------------------------

export const BcfViewpointCreateSchema = z.object({
  guid: z.string().max(36),
  index_in_topic: z.number().default(0),
  camera_type: z.string().max(20),
  camera_view_point: BcfVec3Schema,
  camera_direction: BcfVec3Schema,
  camera_up_vector: BcfVec3Schema,
  field_of_view: z.union([z.number(), z.null()]).optional(),
  field_of_height: z.union([z.number(), z.null()]).optional(),
  components: BcfComponentsSchema.optional(),
  clipping_planes: z.array(BcfClippingPlaneSchema).default([]),
  is_2d: z.boolean().optional().default(false),
  view_state_2d: z
    .object({
      center_x: z.number().default(0),
      center_y: z.number().default(0),
      zoom: z.number().default(1),
      visible_layers: z.array(z.string()).default([]),
      file_type: z.string().default('dxf'),
    })
    .optional(),
  linked_file_id: z.union([z.string().uuid(), z.null()]).optional(),
});

export type BcfViewpointCreateInput = z.infer<typeof BcfViewpointCreateSchema>;

export const BcfTopicCreateSchema = z.object({
  title: z.string().trim().min(1).max(255),
  description: z.union([z.string().max(4000), z.null()]).optional(),
  topic_type: z.string().max(50).default('Issue'),
  topic_status: z.string().max(50).default('Open'),
  priority: z.union([z.string().max(50), z.null()]).optional(),
  stage: z.union([z.string().max(50), z.null()]).optional(),
  assigned_to: z.union([z.string().max(255), z.null()]).optional(),
  labels: z.array(z.string()).default([]),
  due_date: z.union([z.string(), z.null()]).optional(),
  linked_finding_id: z.union([z.string().uuid(), z.null()]).optional(),
  linked_model_id: z.union([z.string().uuid(), z.null()]).optional(),
  viewpoint: BcfViewpointCreateSchema.optional(),
});

export type BcfTopicCreateInput = z.infer<typeof BcfTopicCreateSchema>;

export const BcfTopicUpdateSchema = z.object({
  title: z.string().trim().min(1).max(255).optional(),
  description: z.union([z.string().max(4000), z.null()]).optional(),
  topic_type: z.string().max(50).optional(),
  topic_status: z.string().max(50).optional(),
  priority: z.union([z.string().max(50), z.null()]).optional(),
  stage: z.union([z.string().max(50), z.null()]).optional(),
  assigned_to: z.union([z.string().max(255), z.null()]).optional(),
  labels: z.union([z.array(z.string()), z.null()]).optional(),
  due_date: z.union([z.string(), z.null()]).optional(),
  linked_finding_id: z.union([z.string().uuid(), z.null()]).optional(),
  linked_model_id: z.union([z.string().uuid(), z.null()]).optional(),
});

export type BcfTopicUpdateInput = z.infer<typeof BcfTopicUpdateSchema>;

export const BcfCommentCreateSchema = z.object({
  text: z.string().trim().min(1).max(4000),
  viewpoint_guid: z.union([z.string().max(36), z.null()]).optional(),
});

export type BcfCommentCreateInput = z.infer<typeof BcfCommentCreateSchema>;

// ---------------------------------------------------------------------------
// Import response
// ---------------------------------------------------------------------------

export const BcfImportResponseSchema = z.object({
  imported_count: z.number(),
  topics: z.array(BcfTopicReadSchema),
  warnings: z.array(z.string()),
});

export type BcfImportResponse = z.infer<typeof BcfImportResponseSchema>;

// ---------------------------------------------------------------------------
// Snapshot upload
// ---------------------------------------------------------------------------

export const BcfSnapshotUploadResponseSchema = z.object({
  upload_url: z.string(),
  storage_key: z.string(),
});

export type BcfSnapshotUploadResponse = z.infer<typeof BcfSnapshotUploadResponseSchema>;
